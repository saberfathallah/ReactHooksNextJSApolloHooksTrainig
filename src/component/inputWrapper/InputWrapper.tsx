import React from "react";
import { Formik } from "formik";
import Button from "@material-ui/core/Button";
import Icon from "@material-ui/core/Icon";
import { createStyles, Theme, makeStyles } from "@material-ui/core/styles";
import EditCommentIconButton from "@component/comment/EditCommentIconButton";
import Input from "@component/input";
import validationForm from "@helpers/validation";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    button: {
      margin: theme.spacing(1),
      float: "right",
    },
  })
);
interface CommentEditFormInput {
  description: string;
}
// interface InputWrapperProps  {
//   label: string;
//   setIsClickEdit: (boolean) => void;
//   updateComment: (variables: any) => any;
//   addComment?: any;
//   variables: {
//     description: string;

//   }
// }

const InputWrapper = (props: any) => {
  const { label, addComment, variables, updateComment, setIsClickEdit } = props;
  const initialValues: CommentEditFormInput = variables;
  const classes = useStyles();
  const style = { margin: 8 };

  const onSubmit = async (values) => {
    if (label === "Ajouter un commentaire") {
      await addComment({
        variables: {
          commentInput: {
            ...variables,
            ...values,
          },
        },
      });
    } else {
      if (values.description !== variables.description) {
        await updateComment({
          variables: {
            updateCommentInput: {
              ...variables,
              ...values,
            },
          },
        });
      }
      setIsClickEdit(false);
    }
  }

  return (
    <Formik
      initialValues={initialValues}
      validate={(values): object => validationForm(values)}
      onSubmit={async (values): Promise<void> => onSubmit(values)}
    >
      {({
        values,
        errors,
        touched,
        handleChange,
        handleBlur,
        handleSubmit,
      }): any => (
        <form style={{ display: "contents" }} onSubmit={handleSubmit}>
          <Input
            handleChange={handleChange}
            handleBlur={handleBlur}
            values={values}
            errors={errors}
            touched={touched}
            label={label}
            variables={variables}
            style={style}
          />
          {label === "Ajouter un commentaire" ? (
            <Button
              type="submit"
              variant="contained"
              color="primary"
              size="small"
              className={classes.button}
              endIcon={<Icon>send</Icon>}
            >
              Commenter
            </Button>
          ) : (
            <EditCommentIconButton setIsClickEdit={setIsClickEdit} />
          )}
        </form>
      )}
    </Formik>
  );
};

export default InputWrapper;
