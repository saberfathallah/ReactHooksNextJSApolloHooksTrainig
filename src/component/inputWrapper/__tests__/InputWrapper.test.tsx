import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import { Formik } from "formik";

import InputWrapper from "../InputWrapper";

describe("InputWrapper", () => {
  const setIsClickEdit = jest.fn();
  const updateComment = jest.fn();
  const addComment = jest.fn();

  const props = {
    label: "Ajouter un commentaire",
    setIsClickEdit,
    updateComment,
    addComment,
    variables: {
      description: "description",
    },
  };

  it("should display snapshot", () => {
    const tree = renderer.create(<InputWrapper {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it("should display add comment button", () => {
    const wrapper = mount(<InputWrapper {...props} />);
    expect(wrapper.text()).toContain("Commenter");
  });

  it("should display modify comment icon", () => {
    props.label = "Modifier ton commentaire"

    const wrapper = mount(<InputWrapper {...props} />);

    expect(wrapper.text()).toContain("Modifier ton commentaire");
  });

  it("should update description field on change", () => {
    const wrapper = mount(<InputWrapper {...props} />);
    const descriptionInput = wrapper.find("input");

    descriptionInput.simulate("change", {
      persist: () => {},
      target: {
        name: "description",
        value: "new comment",
      },
    });

    expect(descriptionInput.html()).toMatch("new comment");
  });

  it("should return error for invalid comment", () => {
    const tree = mount(<InputWrapper {...props} />);
    const form = (props: any = { errors: {} }) =>
      tree.find(InputWrapper).find(Formik).renderProp("children")(props);
    const formWithInvalidDescriptionErrors = form({
      values: {
        description: "",
      },
      errors: {
        description: "Required",
      },
      touched: { description: true },
      isSubmitting: false,
    });

    expect(formWithInvalidDescriptionErrors.html()).toMatch(/Required/);
  });

  // it("shouldn't call dislike should call addLike", () => {
  // const addComment = jest.fn();
  //   props.addComment = addComment;

  //   const wrapper = mount(<InputWrapper {...props} />);

  //   //wrapper.find('[type="submit"]').simmulte('click')
  // .find(`input[name="${name}"]`)

  //   wrapper.find('MuiInputBase-root').simmulte('click')
  //   expect(addComment).toHaveBeenCalledTimes(1);
  // });
});
