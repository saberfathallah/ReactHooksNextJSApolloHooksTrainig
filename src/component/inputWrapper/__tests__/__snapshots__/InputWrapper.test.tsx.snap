// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputWrapper should display snapshot 1`] = `
<form
  onSubmit={[Function]}
  style={
    Object {
      "display": "contents",
    }
  }
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    style={
      Object {
        "margin": 8,
      }
    }
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink={true}
      htmlFor="descriptionstandard-full-width"
      id="descriptionstandard-full-width-label"
    >
      Ajouter un commentaire
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        id="descriptionstandard-full-width"
        name="description"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder=""
        required={false}
        type="description"
        value="description"
      />
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-1 MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiButton-label"
    >
      Commenter
      <span
        className="MuiButton-endIcon MuiButton-iconSizeSmall"
      >
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root"
        >
          send
        </span>
      </span>
    </span>
  </button>
</form>
`;

exports[`InputWrapper should display snapshot with Ajouter un commentaire 1`] = `
<form
  onSubmit={[Function]}
  style={
    Object {
      "display": "contents",
    }
  }
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    style={
      Object {
        "margin": 8,
      }
    }
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink={true}
      htmlFor="descriptionstandard-full-width"
      id="descriptionstandard-full-width-label"
    >
      Ajouter un commentaire
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        id="descriptionstandard-full-width"
        name="description"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder=""
        required={false}
        type="description"
        value="description"
      />
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-1 MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiButton-label"
    >
      Commenter
      <span
        className="MuiButton-endIcon MuiButton-iconSizeSmall"
      >
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root"
        >
          send
        </span>
      </span>
    </span>
  </button>
</form>
`;

exports[`InputWrapper should display snapshot with Modifier ton commentaire 1`] = `
<form
  onSubmit={[Function]}
  style={
    Object {
      "display": "contents",
    }
  }
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    style={
      Object {
        "margin": 8,
      }
    }
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
      data-shrink={true}
      htmlFor="descriptionstandard-full-width"
      id="descriptionstandard-full-width-label"
    >
      Modifier ton commentaire
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        id="descriptionstandard-full-width"
        name="description"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder=""
        required={false}
        type="description"
        value="description"
      />
    </div>
  </div>
  <button
    aria-label="modify"
    className="MuiButtonBase-root MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
</form>
`;
